### ISSUE TITLE
[Feature Request] Telegram messenger notifications

### ISSUE THREAD

**Issue by neomvll** [Mikadily/poloniexlendingbot#153]
Can we get a notifier to telegram?
With user ID and API token we can send any message to any telegram bot :)
So, with this feature, we(users) just need put in config file our USER ID and API Token.

**Comment by Evanito** [comment]
Okay, I assume you mean https://telegram.org/
Help me understand why we need this, what kind of notifications would you want to receive?

**Comment by neomvll** [comment]
Nice.
Well, I'm running the lending bot on my RPi and I'm very neurotic with security LOL. So my RPi is all closed for any access from the internet, and I can only access from my home network.
Even if my RPi was open to internet, I think will be great we have some notifications when some actions happens, like:
- when we lending(active loan) some value, we'll get a notification about this loan, like how much in BTC(amount), in % of our total BTC, the loan rate, and what we can't lend yet.
- if any loan is closed and back to our open loan offers, we'll get a notification about how much we are offering now, maybe in % of our total BTC and the amount. As extra, maybe how much we gain in last closed loan? :DDD

I don't know, it is just an idea to we get more fast information about the actions when they happens :)

**Comment by Evanito** [comment]
I also run on a RPi, are you using the webserver feature to see a summary of your bot's activities?

I do entertain the idea of the bot telling you how much you made after a loan closes, that would be a good feature to remind people how much they are actually making. Notifying after every loan opens would be overkill, in my opinion.

However, Telegram does not have huge adoption rates and I fear that this feature will go unused by the vast majority of users. What do you think of it sending an email to notify you of important updates instead? The individual messages could be gathered so that it sends a "digest" at most once an hour (configurable) or so, thus avoiding inbox spam.

**Comment by neomvll** [comment]
I really hope telegram become to a huge adoption, because the features like the bot system is awesome and helpful.
And sure, the important updates is exactly what I mean, I just said what was perfect to me, but important updates for e-mail or telegram notifications(please :DD) will be really big deal!

**Comment by rnevet** [comment]
I'm not sure the bot actually knows now which loans were returned (complete new feature).
About notifications - we could also generate notifications from the web page:
https://developers.google.com/web/fundamentals/engage-and-retain/push-notifications/

But I think a deeper thought about what is the actual content is required (before the channel and frequency are decided).

**Comment by neomvll** [comment]
Well, my idea is to get notifications at any place any time, so telegram would be the best way to my work.
But ok, maybe me and more one or two guys use telegram here and a full feature to we, not worst it, I understand that.
However , any notifications, like this browser notification will be nice too, after all  whatever you doing, the lending bot is awesome :)

**Comment by rnevet** [comment]
@neomvll The main question is what is the expected content of the notifications? on what events are they triggered.

**Comment by neomvll** [comment]
@rnevet I said in my first reply:
" - when we lending(active loan) some value, we'll get a notification about this loan, like how much in BTC(amount), in % of our total BTC, the loan rate, and what we can't lend yet.
 - if any loan is closed and back to our open loan offers, we'll get a notification about how much we are offering now, maybe in % of our total BTC and the amount. As extra, maybe how much we gain in last closed loan? :DDD"

Triggered by active loan starts and active loan closes, respectively.

But as I @Evanito said too:
"What do you think of it sending an email to notify you of important updates instead?"
In this case, I think you know what is important updates more than I :)

Anyway, to me, what I said was the "perfect world", I'll receive the notifications what I need in my telegram anywhere I'm, but ok if to you the complete new feature not worst it, I understand that.

**Comment by laxdog** [comment]
What about:
- Coins are deposited (in exchange account)
- Coins are transferred (exchange to lending)
- Loans are made over xdaysthreshold
- Summary every X minutes (using the botlog.json "last_status")

It could be added as a new log function, `log.sendNotifcations(self, msg='', targets=[], channel='')` for example

It should also be a separate module called Notifications or something. It can then be designed to be extensible to support Telegram, Slack, Email, WhateverElseYouWant.

It will need to use thirdparty libraries, so the module will only be loaded if configured to keep the dependencies at a minimum. 

Probably more important to add the messaging functionality in a useful way, then the events people are interested in can be added at a later date.

**Comment by capitanpiluso** [comment]
If you guys are serious about security, telegram is a bad choice with bad practices.
Signal would be the ideal IM for security purposes.

**Comment by laxdog** [comment]
If it's only posting notifications without sensitive info, I don't think it matters. Though again, I'd be more interested in building the base for a notification / messaging module and letting it grow from there. The platform isn't the concern.

**Comment by rnevet** [comment]
I still don't see the need, just open the html status page from your phone.

**Comment by laxdog** [comment]
My main interest is actually using it through Slack. I'm on Slack all day and I already use it for another bot to give me updates. 

It's an easy way to have persistent continuous logging outside your server and a view of current status without having to open any ports on your server. The same holds true for email updates.

There are definitely benefits to it. Again, I'm more looking at this as a way to introduce a means of notification, I've not fully explored the alerts that would be available. The four I listed are just ones I find interesting. I actually notify on a trade (I sell all XMR mined), transfers (BTC mined / created from trades, then moved to lending) and what is currently the best coin to lend.

![polo](https://cloud.githubusercontent.com/assets/794502/22445461/7653fc52-e73f-11e6-8288-1f7916b35eaf.jpg)

If you don't think it's a great idea, or it's not going to be used much, I'll not bother. I've kind of just been looking for something interesting to contribute. 

The other thing I was considering was the work someone started in https://github.com/Mikadily/poloniexlendingbot/pull/114. That PR seems dead.

**Comment by Evanito** [comment]
@laxdog Any pull request that adds more than it breaks is very likely to get merged. While it may not be a priority for us to code, if any contributor codes Slack/any other IM support I would be willing to work to get it merged.

**Comment by laxdog** [comment]
I've a very basic implementation done using email, slack and telegram. Though everything is hardcoded and I still need to add extra hooks, then all the docs. 

I've kept it pure python too. Since we're only posting message I didn't feel we needed to import a whole library for each message platform. Though that could blow up in my face for debugging!

Is it best to raise the PR now, or wait until it's completely finished? (That might take me a while)
